engineering_tools/carma

```diff
@@ -339,6 +339,174 @@ carma__stop() {
+            exit 1
+        fi
+
+        local IMAGE_NAME="$USERNAME/carma-config:$TAG"
+
+        if [ "$(docker images | grep $USERNAME/carma-config | grep $TAG )" ]; then
+            echo "Found image, applying to carma-config-volume"
+            docker run --rm --name carma-config-data -v carma-config-data:/opt/carma/vehicle/config -d $USERNAME/carma-config:$TAG
+        else
+            echo "Could not find $USERNAME/carma-config:$TAG in local images"
+        fi
+
+
+    elif [ "$1" = "edit-config" ]; then
+        echo "Opening carma-config-data volume shell with read/write privileges..."
+
+        if [ "$(docker volume ls | grep carma-config-data | sed 's/local     //g')" ]; then
+            echo "Found carma-config-data volume"
+        else
+            echo "Missing carma-config-data volume, add with \"carma swarm update-config <image tag>\""
+            exit 1
+        fi
+
+        local carma_base=$(__get_image_from_config carma-base:)
+        if [[ -z $carma_base ]]; then
+            __pull_newest_carma_base
+            carma_base=$(__get_most_recent_carma_base)
+        fi
+
+        docker run -it --rm -v carma-config-data:/opt/carma/vehicle/config $carma_base bash
+
+
+    # Attaches each carma service's STDOUT logs to the current terminal
+    elif [ "$1" = "attach" ]; then
+        echo "Attaching STDOUT logs to terminal..."
+
+        SWARM_SERVICES=$(docker service ls --format "{{.Name}}")
+
+        # this seems to loop forever and cannot be escaped with ctrl+C
+        # performance will likely be poor due to combined log streams
+        for TASK in $SWARM_SERVICES; do
+            docker service logs --follow --details --tail=10 "$TASK" &
+        done
+
+        sleep infinity
+
+
+    # Takes n image tags that are saved locally and pushes them to the registry
+    elif [ "$1" = "set-image" ]; then
+        shift
+        echo "Installing Docker images..."
+
+        if [ "$(docker volume ls | grep carma-registry | sed 's/local     //g')" ]; then
+            echo "Found carma-registry volume"
+        else
+            echo "Missing carma-registry volume, adding now..."
+            docker volume create carma-registry
+        fi
+
+        docker service create --constraint 'node.role==manager' -d --volume "carma-registry:/var/lib/registry" -p 5000:5000 --name carma-registry-service registry:2
+        
+        IMAGE_NAMES="$*"
+
+        for SOURCE_IMAGE_NAME in $IMAGE_NAMES
+        do
+        
+        echo "Pushing $SOURCE_IMAGE_NAME to localhost:5000"
+
+        docker tag "$SOURCE_IMAGE_NAME" "localhost:5000/$SOURCE_IMAGE_NAME"
+        docker push "localhost:5000/$SOURCE_IMAGE_NAME"
+        docker rmi "localhost:5000/$SOURCE_IMAGE_NAME"
+
+        done
+
+
+    elif [ "$1" = "down" ]; then
+        echo "Shutting down swarm..."
+
+        if [ "$(docker node ls | grep worker )" ]; then
+            ssh dev@192.168.88.101 "docker swarm leave"
+        fi
+        docker swarm leave --force
+
+
```

Comment:
Another command I think we'd like to support is `carma swarm config-status` which will be similar to `carma-config__status()`, and will printout the contents of each docker-compose file currently set. This will require that carma-config-data has been set by the user with `carma swarm update-config`.